src/Data/JSONPath/Parser.hs

```diff
@@ -203,6 +203,46 @@ comparable endParser = do
     <|> CmpBool <$> bool
     <|> CmpNull <$ string "null"
     <|> CmpPath <$> singularPath endParser
+    <|> CmpFun <$> functionExpr endParser
+
+functionExpr :: Parser a -> Parser FunctionExpr
+functionExpr endParser = do
+  FunctionExpr
+    <$> functionName
+    <*> inParens (functionArgs endParser)
+
+functionArgs :: Parser a -> Parser [FunctionArgument]
+functionArgs endParser = do
+  firstArg <- functionArgument endParser
+  restArgs <- many (char ',' *> functionArgument endParser)
+  pure (firstArg : restArgs)
+
+functionArgument :: Parser a -> Parser FunctionArgument
+functionArgument endParser =
+  (ArgLiteral <$> try (comparable endParser)) -- TODO: need to prohibit CmpPath or CmpFun
+    <|> (ArgFilterQuery <$> filterQuery')
+    <|> (ArgLogicalExpr <$> filterExpr endParser)
+    <|> ArgFunctionExpr <$> functionExpr endParser
+
+-- filterQuery' is like filterQuery except it can parse a prefix of the input
+-- TODO: unify filterQuery' and filterQuery
+filterQuery' :: Parser FilterQuery
+filterQuery' =
+  FilterQuery <$> beginningPoint <*> jsonPath'
+
+-- jsonPath' is like jsonPath except it can parse a prefix of the input
+-- TODO: unify jsonPath' and jsonPath
+jsonPath' :: Parser [JSONPathElement]
+jsonPath' = do
+  some jsonPathElement
+
+functionName :: Parser FunctionName
+functionName =
+  FunLength <$ string "length"
+    <|> FunCount <$ string "count"
+    <|> FunMatch <$ string "match"
+    <|> FunSearch <$ string "search"
+    <|> FunValue <$ string "value"
```

Comment:
I think this might blow in your face in the future if any two strings have common prefixes.

Reply:
Thanks @ffaf1. I'll document it for now. Seems premature to add `try` until we need it. As you point out, this will become apparent pretty quickly once it crops up and a test will easily catch it.

Reply:
Another idea could be to parse any function name according to the [spec](https://ietf-wg-jsonpath.github.io/draft-ietf-jsonpath-base/draft-ietf-jsonpath-base.html#name-function-extensions) and then pattern match over the valid ones. I think this would generate better errors like `"foo" is not a known function`.

Reply:
I meant you can check validity while parsing. So, we can keep a sum type. Once we know which function we parsed, we can parse the required number of arguments.

Reply:
That's fair and fits in with the [parse, don't validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/) approach. However, I think it would be helpful to check well-formedness before well-typdeness in the spirit of the IETF spec. Then basic syntax errors will be diagnosed before, and differently to, type errors.

What I might try is a kind of two-pass parser (see "Don’t be afraid to parse data in multiple passes" in the linked article) where I used Megaparsec to do the well-formdness check and produce an intermediate type for function expressions (see "Keep denormalized representations of data behind abstraction boundaries" in trhe linked article) which can then be checked for well-typedness to produce a type along the lines you suggest. I'm not yet sure whether this will work or how it will look, but I'd like to take a crack at it.